### PR TITLE
Teleportation

### PR DESCRIPTION
**Pull request by imgntn**
This PR adds teleportation to the array of movement options.

To test:
1. Download this build and reload your default scripts.  You should see teleport.js in the defaults now.

To teleport: 
1. Press and hold the very middle of the Vive controller thumbpad  and the trigger to activate teleportation.  Or the A/X button on the Oculus Touch controller.  Or the 1 button on the Hydras.
2. Point the target toward where you'd like to go.
3. Release either the thumb or the trigger.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: f4d409f1e16b56bc08558edca6fa1928647b11c1 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ff745a345e17e3ba44dc5e7925338f252d9538fe available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 056e9e3a720851dd0635e3315df2be861481b3fb available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cbd1f8df88d83d8ea3c703b9d50b214b2967d751 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c78dbe26b65fe19c96644e00221a09c8889dccb2 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 26950eaaa6efed9de1a7a9ce4495d943108210fa available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b0494ec9d53d194637b78dcdf172c85507196434 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

File: scripts/system/controllers/teleport.js [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+var FADE_IN_INTERVAL = 50;
+var FADE_OUT_INTERVAL = 50;
+
+var BEAM_MODEL_URL = "http://hifi-content.s3.amazonaws.com/james/teleporter/teleportBeam.fbx";
+var BEAM_MODEL_URL_NO_INTERSECTION = "http://hifi-content.s3.amazonaws.com/james/teleporter/teleportBeam2.fbx";
+
+
+var STRETCHY_BEAM_DIMENSIONS_X = 0.04;
+var STRETCHY_BEAM_DIMENSIONS_Y = 0.04;
+var STRETCHY_BEAM_DIMENSIONS_Z_NO_INTESECTION = 20;
```
Comment (PhilipRosedale) [inline scripts/system/controllers/teleport.js]: spelling

File: scripts/system/controllers/teleport.js [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+//     y: 0.93,
+//     z: 1.22
+// };
+
+function ThumbPad(hand) {
+    this.hand = hand;
+    var _this = this;
+
+    this.buttonPress = function(value) {
+        print('jbp pad press: ' + value + " on: " + _this.hand)
```
Comment (PhilipRosedale) [inline scripts/system/controllers/teleport.js]: debug

File: scripts/system/controllers/teleport.js [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+    var _this = this;
+    this.intersection = null;
+    this.targetProps = null;
+    this.targetOverlay = null;
+    this.stretchyBeam = null;
+    this.noIntersectionStretchyBeam = null;
+    this.updateConnected = null;
+
+    this.initialize = function() {
+        print('jbp initialize')
```
Comment (PhilipRosedale) [inline scripts/system/controllers/teleport.js]: debug

File: scripts/system/controllers/teleport.js [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+            dimensions: TARGET_MODEL_DIMENSIONS,
+            visible: true,
+        };
+
+        _this.targetOverlay = Overlays.addOverlay("model", targetOverlayProps);
+    };
+
+
+    this.createMappings = function() {
+        print('jbp create mappings internal');
```
Comment (PhilipRosedale) [inline scripts/system/controllers/teleport.js]: debug print

File: scripts/system/controllers/teleport.js [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+        print('jbp disable mappings internal')
+        Controller.disableMapping(teleporter.telporterMappingInternalName);
+    };
+
+    this.enterTeleportMode = function(hand) {
+        if (inTeleportMode === true) {
+            return;
+        }
+
+        print('jbp hand on entering teleport mode: ' + hand);
```
Comment (PhilipRosedale) [inline scripts/system/controllers/teleport.js]: debug print

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2d073cc99c2b50b042c69af331989efd82bbc774 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b2ef491b973ff704567180bbfae22ef94805a027 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e993a695195aca298f1898449d2cd0ed0a16073b available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 696169ed3731f93d95b104c862e34313d259e488 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1728878982d165b2c232c73d7e3402e98e465fd5 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5626938fb40b97c59ff130d61a8a59bf9c5481d0 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 54e0c131a562474b38968041d380d4ee9b0a40c8 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3c3bad4d3bbb1683cf7b698f31a3aeb92fd6481f available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bd3fa60927eff9bc4e41dff21038426feed79a81 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 93081e73d8b784276186927f6bee65db7eeaf45a available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: be770c81777dea43f48f149ad2c9636bb8053f87 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4a9f6e677c4187bb89c23b227409630ac203a987 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by imgntn** [comment]
testing links for ryan

heres one with 100ms delay  for vive https://rawgit.com/imgntn/hifi/8212ada4dec4930e7b270b8ed0a342cd0e43fc73/scripts/system/controllers/teleport.js

[12:44]  
here's one with the trigger + thumb https://rawgit.com/imgntn/hifi/93081e73d8b784276186927f6bee65db7eeaf45a/scripts/system/controllers/teleport.js

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5c83db3b28302a817363877780dbde73dff2e574 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1a463665579c5a438c15059e8a3586468050f3e8 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f8ff0da9013515e8a1623e3ae09f4fbdae407ca9 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 24832df14a7d7ce2fc7c9324c5a2de4c51295194 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3492367c63e27c544e900b043b159399d3b9c02e available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 38e580b211908698f155dec7ab62035fbf6d6904 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4f899d875288e0b5e9b3dd8af4b4e14c8e242e3c available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ee970c5e07a6c5ace193951a6f6f69ceb52df53a available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5ca14b528bf7e5260e819517b0c24468e83d0da2 available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6249f3ed12ea4c67238ca53529bb0d480312773d available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190.dmg

**Comment by RyanDowne** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;Yka588mS&#x2F;63-teleporting"> Teleporting</a>

**Comment by kevinmthomasse** [comment]
testing

**Comment by kevinmthomasse** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**:  ```teleport.js``` was present in the defaults as I was successfully able to test using a set of hydras to which upon pressing the trigger showed a beam to which I was able to successfully press the 1 key to which it teleported my avatar to the beam location.

**Comment by hifi-gustavo** [comment]
Build console output for commit: dc9c06a159810f0f3fdc039fd9896c1f172042ac available at: https://gist.github.com/fbc249454996439e81ce30a70b33edb6<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190-dc9c06a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8190-dc9c06a.dmg

**Pull request merged**